kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/parquet/ParquetSchemaUtils.java

```diff
@@ -265,38 +272,60 @@ private static Type toParquetType(
         return type;
     }
 
-    private static Type toParquetArrayType(ArrayType arrayType, String name, Repetition rep) {
-        // We will be supporting the 3-level array structure only. 2-level array structure will
-        // be supported in the future.
+    private static Type toParquetArrayType(
+            StructField nearestAncestor, // nearest ancestor with a StructField
+            String relativePath, // relative path to this element from the nearestAncestor
+            ArrayType arrayType,
+            String name,
+            Repetition rep) {
+        // We will be supporting the 3-level array structure only. 2-level array structures are
+        // a very old legacy versions of Parquet which Kernel doesn't support writing as.
+
+        String elementRelativePath = relativePath + ".element";
         return Types
                 .buildGroup(rep)
                 .as(LogicalTypeAnnotation.listType())
                 .addField(Types.repeatedGroup()
                         .addField(toParquetType(
+                                nearestAncestor,
+                                elementRelativePath,
                                 arrayType.getElementType(),
                                 "element", /* name */
                                 arrayType.containsNull() ? OPTIONAL : REQUIRED,
-                                Optional.empty()))
+                                getNestedFieldId(nearestAncestor, elementRelativePath)))
```

Comment:
what happens if the result of `getNestedFieldId(nearestAncestor, elementRelativePath)` is `Optional.empty()`? what does that mean?

Reply:
I have a follow up PR that adds validation methods. Will open shortly. It looks like there is an implicit assumption in the protocol that
1) either all columns have field id or none
2) similar to 1, but for the nested field id: all of them have or none of them.

I will add the validation before we go to the conversion to Parquet type logic, so that this shouldn't be a problem.